### ISSUE TITLE
[Documentation] Document tensor typing

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#98]
### Discussed in https://github.com/beartype/beartype/discussions/96

<div type='discussions-op-text'>

<sup>Originally posted by **amitkparekh** February  7, 2022</sup>
I love the idea of this project and I _want_ to use beartype in all my projects going forwards. I've been lurking for a while and I don't think it's been suggested so here's my question/request: 

How can I use this with [torch's Tensor](https://pytorch.org/docs/stable/tensors.html)? 

I'm writing a lot of ML/NLP research code and I want to use this in my projects to not need to keep typing the shape in the comments, and so I can just further trust that each tensor is what I've told it to be. 

I know you support numpy OOTB, but I'm not sure whether it can translate directly to torch, and whether it can follow the various reshaping methods that make square tensors fit in round-shaped holes. 

I've had trouble fully grasping how to do things with beartype because I have trouble understanding the README. Don't get me wrong, I absolutely _love_ the writing, but I genuinely have trouble understanding it because it's just a lot of information that adds more information super fast.

</div>

**Comment by leycec** [comment]
> ...because I have trouble understanding the README.

**You are not alone.** We also have trouble understanding 1,613-page doctoral theses in obscure LaTeX3.1415 formats littered with unpronounceable acronyms, jargon-on-jargon, and recursive puns – which is why I let my wife read those. In this case, beartype usage could best be synopsized with this fist-pumping motto:

* **[`beartype.vale.Is[lambda obj: ...]`](https://github.com/beartype/beartype#validator-api) is QA King.**

[`beartype.vale.Is[lambda obj: ...]`](https://github.com/beartype/beartype#validator-api) is our functional validator API. It's what you do when all of the following apply:

* **What you want to do isn't covered by an existing PEP standard.** Almost certain that's the case here.
* **You don't want to dive *too* deep into beartype validators.** Efficient alternatives to our functional validator API do exist – but they're *much* less general-purpose. For most purposes (almost certainly including this one), the functional validator API suffices. Always go there first.
* **You know exactly how, what, and why you want to validate.** In this case, "How can I use this with [torch's Tensor](https://pytorch.org/docs/stable/tensors.html)?" is a *really* broad question. To preserve dwindling sanity, let's pretend you instead asked:

> "How can I type-check a PyTorch tensor to have the expected dtype?"

Apropos question, Amit. All of the above conditions now apply! We can now type you up a possibly working minimal-length example (MLE) combining the fearsome powers of @beartype **+** PyTorch **+** [PEP 593](https://www.python.org/dev/peps/pep-0593) to validate callable parameters and returns as strictly floating-point *or* integral PyTorch tensors:

``` python
from beartype import beartype
from beartype.vale import Is
from torch import (
    float as torch_float,
    int as torch_int,
    tensor,
)
from typing import Annotated   # <--------------- if Python ≥ 3.9.0
# from typing_extensions import Annotated   # <-- if Python < 3.9.0

# PEP-compliant type hint matching only a floating-point PyTorch tensor.
TorchTensorFloat = Annotated[tensor, Is[
    lambda tens: tens.type() is torch_float]]

# PEP-compliant type hint matching only an integral PyTorch tensor.
TorchTensorInt = Annotated[tensor, Is[
    lambda tens: tens.type() is torch_int]]

# Type-check everything like an NLP babelfish.
@beartype
def deep_dream(dreamy_tensor: TorchTensorFloat) -> TorchTensorInt:
    return dreamy_tensor.type(dtype=torch_int)
```

Of course, `beartype.vale.Is[...]` supports arbitrarily complex Turing-complete Python expressions. The above MLE generalizes to typing the dimensionality, device, and other metadata of a tensor to whatever degree of specificity you need.

`beartype.vale.Is[...]`: *it's lambdas all the way down.*

**Comment by leycec** [comment]
Actually, let's expand this issue to cover *all* popular Python tensor frameworks. Surprisingly, this is (mostly) easier done than said. Some frameworks now provide official runtime type-checking support. This includes:

* Google's [`jaxtyping`](https://github.com/google/jaxtyping), publishing JAX-specific type hints supporting both runtime and static type-checking.
* [*Insert other packages here.*]

See also [this related issue at the `jaxtyping` issue tracker](https://github.com/google/jaxtyping/issues/38).

**Comment by leycec** [comment]
**Resolved by d6faaac.** Another doco request bites the dust. Thanks so much for the initial impetus, @amitkparekh. I hope life in the research trenches finds you well, wherever you may now roam. :scotland:
